tests/drivers/build_all/sensor/src/generic_test.c

```diff
@@ -0,0 +1,186 @@
+#include <zephyr/ztest.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/rtio/rtio.h>
+
+/*
+ * Set up an RTIO context that can be shared for all sensors
+ */
+
+static enum sensor_channel iodev_all_channels[SENSOR_CHAN_ALL];
+static struct sensor_read_config iodev_read_config = {
+	.sensor = NULL,
+	.channels = iodev_all_channels,
+	.count = 0,
+	.max = SENSOR_CHAN_ALL,
+};
+RTIO_IODEV_DEFINE(iodev_read, &__sensor_iodev_api, &iodev_read_config);
+
+/* Create the RTIO context to service the reading */
+RTIO_DEFINE_WITH_MEMPOOL(sensor_read_rtio_ctx, 1, 1, 1, 64, 4);
+
+/** Number of test points to use in the generic test. These are automatically generated. */
+#define NUM_EXPECTED_VALS (5)
+
+/**
+ * @brief Helper function the carries out the generic sensor test for a given sensor device.
+ *        Verifies that the device has a suitable emulator that implements the backend API and
+ *        skips the test gracefully if not.
+ */
+static void run_generic_test(const struct device *dev)
+{
+	zassert_not_null(dev, "Cannot get device pointer. Is this driver properly instantiated?");
+
+	const struct emul *emul = emul_get_binding(dev->name);
+
+	/* Skip this sensor if there is no emulator loaded. */
+	if (!emul) {
+		ztest_test_skip();
+	}
+
+	/* Also skip if this emulator does not implement the backend API. */
+	if (!emul_sensor_backend_is_supported(emul)) {
+		ztest_test_skip();
+	}
+
+	/*
+	 * Begin the actual test sequence
+	 */
+
+	static struct {
+		bool supported;
+		bool received;
+		q31_t expected_values[NUM_EXPECTED_VALS];
+		q31_t epsilon;
+		int8_t expected_value_shift;
+	} channel_table[SENSOR_CHAN_ALL];
+
+	/* Discover supported channels on this device and fill out our sensor read request */
+	for (enum sensor_channel ch = 0; ch < ARRAY_SIZE(channel_table); ch++) {
+		if (SENSOR_CHANNEL_3_AXIS(ch)) {
+			continue;
+		}
+
+		q31_t lower, upper;
+		int8_t shift;
+
+		if (emul_sensor_backend_get_sample_range(emul, ch, &lower, &upper,
+							 &channel_table[ch].epsilon, &shift) == 0) {
+			/* This channel is supported */
+			channel_table[ch].supported = true;
+
+			/* Add to the list of channels to read */
+			iodev_all_channels[iodev_read_config.count++] = ch;
+
+			/* Generate a set of NUM_EXPECTED_VALS test values */
+			channel_table[ch].expected_value_shift = shift;
+			for (size_t i = 0; i < NUM_EXPECTED_VALS; i++) {
+				channel_table[ch].expected_values[i] =
+					lower + (i * (upper - lower) / (NUM_EXPECTED_VALS - 1));
+			}
+		}
+	}
+	iodev_read_config.sensor = dev;
+
+	/* Do a read of all channels for quantity NUM_EXPECTED_VALS rounds. */
+
+	int rv;
+	uint8_t *buf = NULL;
+	uint32_t buf_len = 0;
+	const struct sensor_decoder_api *decoder;
+	sensor_frame_iterator_t fit;
+	sensor_channel_iterator_t cit;
+	enum sensor_channel channel;
+	q31_t q;
+	int8_t shift;
+	int missing_channel_count;
+
+	zassert_ok(sensor_get_decoder(dev, &decoder));
+
+	for (size_t round = 0; round < NUM_EXPECTED_VALS; round++) {
```

Comment:
Nit: Since "rounding" is also meaningful in a sensor context, I might suggest "iteration."